infra/Pos/Reporting/Methods.hs

```diff
@@ -61,6 +72,125 @@ import           Pos.Reporting.MemState (HasLoggerConfig (..), HasReportServers
+    e  <- try $ liftIO $ httpLbs rq reportManager
+    whenLeft e $ \(e' :: SomeException) -> throwM $ SendingError e'
+  where
+    partFile' fp = Form.partFile (toFileName fp) fp
+    toFileName = toText . takeFileName
+    mkReportInfo curTime =
+        ReportInfo
+        { rApplication = appName
+        -- We are using version of 'cardano-sl-infra' here. We agreed
+        -- that the version of 'cardano-sl' and it subpackages should
+        -- be same.
+        , rVersion = version
+        , rBuild = pretty compileInfo
+        , rOS = toText (os <> "-" <> arch)
+        , rMagic = getProtocolMagic protocolMagic
+        , rDate = curTime
+        , rReportType = reportType
+        }
+
+
+-- | Given logger config, retrieves all (logger name, filepath) for
+-- every logger that has file handle. Filepath inside does __not__
+-- contain the common logger config prefix.
+retrieveLogFiles :: LoggerConfig -> [([Text], FilePath)]
+retrieveLogFiles lconfig = fromLogTree $ lconfig ^. lcTree
+  where
+    fromLogTree lt =
+        let curElems = map ([],) (lt ^.. ltFiles . each . hwFilePath)
+            addFoo (part, node) = map (first (part :)) $ fromLogTree node
+        in curElems ++ concatMap addFoo (lt ^. ltSubloggers . to HM.toList)
+
+-- | Pass a list of absolute paths to log files. This function will
+-- archive and compress these files and put resulting file into log
+-- directory (returning filepath is absolute).
+compressLogs :: (MonadIO m) => [FilePath] -> m FilePath
+compressLogs files = liftIO $ do
+    tar <- tarPackIndependently files
+    tarxz <-
+        BS.concat <$>
+        runConduitRes (yield tar .| Lzma.compress (Just 0) .| consume)
+    aName <- getArchiveName
+    withFile aName WriteMode $ \handle -> do
+        BS.hPut handle tarxz
+        hFlush handle
+    pure aName
+  where
+    tarPackIndependently :: [FilePath] -> IO ByteString
+    tarPackIndependently paths = do
+        entries <- forM paths $ \p -> do
+            unlessM (doesFileExist p) $ throwM $
+                PackingError $ "can't pack log file " <> fromString p <>
+                               " because it doesn't exist or it's not a file"
+            tPath <- either (throwM . PackingError . fromString)
+                            pure
+                            (Tar.toTarPath False $ takeFileName p)
+            pabs <- canonicalizePath p
+            Tar.packFileEntry pabs tPath
+        pure $ BSL.toStrict $ Tar.write entries
+    getArchiveName = liftIO $ do
+        curTime <- formatTime defaultTimeLocale "%q" <$> getCurrentTime
+        pure $ "report-" <> curTime <> ".tar.lzma"
+
+-- | Creates a temp file from given text
+withTempLogFile :: (MonadIO m, MonadMask m) => Text -> (FilePath -> m a) -> m a
+withTempLogFile rawLogs action = do
+    withSystemTempFile "main.log" $ \tempFp tempHandle -> do
+        liftIO $ do
+            TIO.hPutStrLn tempHandle rawLogs
+            hClose tempHandle
```

Comment:
Is it safe to close the handler twice?

Reply:
If you don't know for sure I would recommend `hFlush` instead

Reply:
Yes, it is safe to call it twice.

Reply:
https://www.stackage.org/haddock/lts-9.9/base-4.9.1.0/System-IO.html#v:hClose

>>>Performing hClose on a handle that has already been closed has no effect; doing so is not an error.